src/storage/txn/scheduler.rs

```diff
@@ -790,8 +791,17 @@ impl<E: Engine, L: LockManager> Scheduler<E, L> {
     }
 
     fn on_release_locks(&self, released_locks: ReleasedLocks) {
-        let mut legacy_wake_up_list = vec![];
-        let mut delay_wake_up_futures = vec![];
+        // This function is always called when holding the latch of the involved keys.
+        // So if we found the lock waiting queues are empty, there's no chance
+        // that other threads/commands adds new lock-wait entries to the keys
+        // concurrently. Therefore it's safe to skip waking up when we found the
+        // lock waiting queues are empty.
+        if self.inner.lock_wait_queues.is_empty() {
```

Comment:
Can we skip waking up according to the length of `queue_map`?

Reply:
As you said it is most important to avoid the high pool work. Then, can we just remove this empty check, or is it equally important?

Reply:
It's not as important. I did this just because I noticed it seems to be able to be optimized while I'm investigating the regression problem. I think it's better if in most cases we can check only an atomic value instead of lock and check the concurrent hashmap every time. I don't have any idea if it can be removed since I didn't test it... so I suggest keep this optimization.

Reply:
OK, I got it. My thought is that this optimization may not always be so effective. As long as there are some lock waiting, although they are irrelavant to this key, the optimization does not work.

Wrapping the atomic int into a prometheus observer can avoid the duplication, but it's a bit complex so I'm not advising this way.